In [1]:
data = [1, 4, 9, 15, 3, 20, 32, 11, 29]
k = 2
centroids = [1, 4]  # 初始中心
history = []  # 记录迭代历史

def format_cluster(cluster):
    return "{" + ",".join(map(str, sorted(cluster))) + "}"

iteration = 1
while True:
    # 分配簇
    clusters = {0: [], 1: []}
    assignments = []
    
    # 记录当前中心
    current_m1, current_m2 = centroids
    
    # 分配过程
    for point in data:
        d1 = abs(point - current_m1)
        d2 = abs(point - current_m2)
        
        # 处理平局情况（根据第3次迭代的示例）
        if d1 == d2:
            cluster = 1  # 平局时分配到K2
            reason = f"平局（距离{d1} = {d2}，假设保留在K2）"
        else:
            cluster = 0 if d1 < d2 else 1
            reason = f"距离{min(d1, d2)} < {max(d1, d2)}"
        
        clusters[cluster].append(point)
        assignments.append(f"{point} → K{cluster+1}（{reason}）")
    
    # 记录当前迭代结果
    history.append({
        "iteration": iteration,
        "centroids": (current_m1, current_m2),
        "clusters": clusters,
        "assignments": assignments
    })
    
    # 计算新中心
    new_m1 = round(sum(clusters[0])/len(clusters[0]), 4) if clusters[0] else current_m1
    new_m2 = round(sum(clusters[1])/len(clusters[1]), 4) if clusters[1] else current_m2
    
    # 检查收敛
    if (new_m1, new_m2) == (current_m1, current_m2):
        break
    
    centroids = [new_m1, new_m2]
    iteration += 1

# 打印完整过程
for entry in history:
    print(f"\n---\n### 第{entry['iteration']}次迭代")
    print(f"**当前中心**:  K1={entry['centroids'][0]:g}, K2={entry['centroids'][1]:g}")
    
    print("\n**分配过程**:")
    for desc in entry['assignments']:
        print(f"- {desc}")
    
    print("\n**簇分配结果**:")
    print(f"- K1 = {format_cluster(entry['clusters'][0])}")
    print(f"- K2 = {format_cluster(entry['clusters'][1])}")
    
    # 计算新中心（显示精确计算过程）
    if entry['clusters'][0]:
        k1_sum = sum(entry['clusters'][0])
        k1_len = len(entry['clusters'][0])
        k1_center = f"({'+'.join(map(str, entry['clusters'][0]))})/{k1_len} = {k1_sum/k1_len:.4f}"
    else:
        k1_center = "保持原中心"
    
    if entry['clusters'][1]:
        k2_sum = sum(entry['clusters'][1])
        k2_len = len(entry['clusters'][1])
        k2_center = f"({'+'.join(map(str, entry['clusters'][1]))})/{k2_len} = {k2_sum/k2_len:.4f}"
    else:
        k2_center = "保持原中心"
    
    print("\n**更新中心**:")
    print(f"- K1新中心：{k1_center}")
    print(f"- K2新中心：{k2_center}")

print("\n---\n### 最终聚类结果")
final_clusters = history[-1]['clusters']
print(f"- K1: {format_cluster(final_clusters[0])}")
print(f"- K2: {format_cluster(final_clusters[1])}")
print("**收敛条件满足，算法终止**")


---
### 第1次迭代
**当前中心**:  K1=1, K2=4

**分配过程**:
- 1 → K1（距离0 < 3）
- 4 → K2（距离0 < 3）
- 9 → K2（距离5 < 8）
- 15 → K2（距离11 < 14）
- 3 → K2（距离1 < 2）
- 20 → K2（距离16 < 19）
- 32 → K2（距离28 < 31）
- 11 → K2（距离7 < 10）
- 29 → K2（距离25 < 28）

**簇分配结果**:
- K1 = {1}
- K2 = {3,4,9,11,15,20,29,32}

**更新中心**:
- K1新中心：(1)/1 = 1.0000
- K2新中心：(4+9+15+3+20+32+11+29)/8 = 15.3750

---
### 第2次迭代
**当前中心**:  K1=1, K2=15.375

**分配过程**:
- 1 → K1（距离0.0 < 14.375）
- 4 → K1（距离3.0 < 11.375）
- 9 → K2（距离6.375 < 8.0）
- 15 → K2（距离0.375 < 14.0）
- 3 → K1（距离2.0 < 12.375）
- 20 → K2（距离4.625 < 19.0）
- 32 → K2（距离16.625 < 31.0）
- 11 → K2（距离4.375 < 10.0）
- 29 → K2（距离13.625 < 28.0）

**簇分配结果**:
- K1 = {1,3,4}
- K2 = {9,11,15,20,29,32}

**更新中心**:
- K1新中心：(1+4+3)/3 = 2.6667
- K2新中心：(9+15+20+32+11+29)/6 = 19.3333

---
### 第3次迭代
**当前中心**:  K1=2.6667, K2=19.3333

**分配过程**:
- 1 → K1（距离1.6667 < 18.3333）
- 4 → K1（距离1.3333 < 15.333300000000001）
- 9 → K1（距离6.3332999999999995 < 10.333300000000001）
- 15 → K2（距离4.333300000000001 < 12.3333）
- 3 → K1（距离

### 第1次迭代
**当前中心**:  K1=1, K2=4
**簇分配结果**:
- K1 = {1}
- K2 = {3,4,9,11,15,20,29,32}
**更新中心**:
- K1新中心：(1)/1 = 1.0000
- K2新中心：(4+9+15+3+20+32+11+29)/8 = 15.3750

### 第2次迭代
**当前中心**:  K1=1, K2=15.375
**簇分配结果**:
- K1 = {1,3,4}
- K2 = {9,11,15,20,29,32}
**更新中心**:
- K1新中心：(1+4+3)/3 = 2.6667
- K2新中心：(9+15+20+32+11+29)/6 = 19.3333

### 第3次迭代
**当前中心**:  K1=2.6667, K2=19.3333
**簇分配结果**:
- K1 = {1,3,4,9,11}
- K2 = {15,20,29,32}
**更新中心**:
- K1新中心：(1+4+9+3+11)/5 = 5.6000
- K2新中心：(15+20+32+29)/4 = 24.0000

### 第4次迭代
**当前中心**:  K1=5.6, K2=24
**簇分配结果**:
- K1 = {1,3,4,9,11}
- K2 = {15,20,29,32}
**更新中心**:
- K1新中心：(1+4+9+3+11)/5 = 5.6000
- K2新中心：(15+20+32+29)/4 = 24.0000

### 最终聚类结果
- K1: {1,3,4,9,11}
- K2: {15,20,29,32}
**收敛条件满足，算法终止**